### PR TITLE
Route method control and CORS

### PR DESCRIPTION
**Pull request by boecklic**
This PR resolves issue #3071 and replaces the old PR #2803. PR #2803 has a lot of unresolved discussions. Currently the PR just copy-paste changes as is in the old PR, so when reviewing this, have a quick look at the discussions in PR #2803. 

Edit by  @procrastinatio 
:bomb: Please carefully test, it may break many things. Most of the content of this PR is however useless, as most headers are rewritten by apache.

Test link = http://mf-chsdi3.int.bgdi.ch/route_method_control_clean/

https://map.geo.admin.ch/?api_url=//mf-chsdi3.int.bgdi.ch/route_method_control_clean

### CODE REVIEW

File: chsdi/__init__.py [hunk 1]
```diff
@@ -70,61 +74,277 @@ def main(global_config, **settings):
...
-    config.add_route('qrcodegenerator', '/qrcodegenerator')
-    config.add_route('sitemap', '/sitemap')
-    config.add_route('luftbilder', '/luftbilder/viewer.html')
-    config.add_route('historicalmaps', '/historicalmaps/viewer.html')
-    config.add_route('checker', '/checker')
-    config.add_route('checker_dev', '/checker_dev')
-    config.add_route('downloadkml', '/downloadkml')
-
-    config.add_route('files_collection', '/files', request_method=('GET', 'POST', 'DELETE'))
+    add_cors_route(config, '/rest/services', 'topics', methods=['GET'])
```
Comment (boecklic) [inline chsdi/__init__.py]: Adding the CORS routes imho adds a lot of duplicated code, maybe it's worth packing the two into a generic method that adds normal and cors routes and makes it easier to see which routes are defined how?

File: requirements.txt [hunk 2]
```diff
@@ -32,7 +32,7 @@ pycodestyle==2.0.0
 pyflakes==1.2.3
 Pygments==2.1.3
 pyproj==1.9.5.1
-pyramid==1.8.3
+pyramid==1.9.1
```
Comment (boecklic) [inline requirements.txt]: Is this upgrade save?

Comment (procrastinatio) [reply requirements.txt]: Use my PR https://github.com/geoadmin/mf-chsdi3/pull/3086 (Pyramid 1.10)

File: tests/integration/test_admin.py [hunk 3]
```diff
@@ -8,3 +8,7 @@ class TestAdminkml(TestsBase):
     def test_admin_kml(self):
         resp = self.testapp.get('/admin/kml', status=200)
         self.assertEqual(resp.content_type, 'text/html')
+
+    def test_admin_kml_post(self):
+        resp = self.testapp.post('/admin/kml', '', status=404)
+        self.assertEqual(resp.content_type, 'text/plain')
```
Comment (procrastinatio) [inline tests/integration/test_admin.py]: Is it OK? Normally, all errors are returned as JSON

File: chsdi/__init__.py (hunk 1 above)
Comment (procrastinatio) [reply chsdi/__init__.py]: Hmm, https://github.com/geoadmin/mf-chsdi3/blob/d1ee3ff291d5b974ae6cb1ff7a8f0dbcd4240af8/chsdi/__init__.py#L25

**Comment by pakb** [comment]
@procrastinatio would this be ready for Wednesday? Or do we add this for next milestone?

**Comment by pakb** [comment]
what's the status on this?

**Comment by procrastinatio** [comment]
This is a bit cleaner, but I really don't like this PR, because we have now routing and CORS at three different levels: `varnish`, `apache` and `pyramid`, and this is not easy to clean up as `route` are so entangled.

**Comment by danduk82** [comment]
@davidoesch, we could add some kind of "deprecated warning" list on API docu before merging this, what do you think?

**Comment by davidoesch** [comment]
users don't read the API docu.
So: use KIBANA and analyse which hosts /Domains / Applications are affected / we will break with this fix. 
Then we will inform them and give them a 3 month notice time to  react on changes

**Comment by danduk82** [comment]
this will break a lot of clients...

**Comment by ltshb** [comment]
@boecklic can we close this ?

**Comment by procrastinatio** [comment]
I guess it has been superceded by https://github.com/geoadmin/mf-chsdi3/pull/3987 and may be deleted.
Any concern @ltshb @boecklic  @hansmannj @ltclm @rebert ?

**Comment by boecklic** [comment]
Jup

**Pull request closed without merging**